### PR TITLE
ondir: init at 0.2.3

### PR DESCRIPTION
**Pull request by michaelCTS**
###### Description of changes

[ondir](https://github.com/alecthomas/ondir/) is an alternative to [direnv](https://direnv.net/) that allows executing commands on directory entry and exit. Unlike `direnv`, `ondir` doesn't execute the command or script in a subshell, but inserts the config into the running shell.  
This is done by the user manually overriding shell builtins `cd`, `pushd`, and `popd`. 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/ondir/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  version = "0.2.2";
+
+  src = fetchFromGitHub {
+    owner = "alecthomas";
+    repo = "ondir";
+    rev = "55279f0366c32e4295c9f114cbee5ac7776e4d53";
+    hash = "sha256-Ds2raQv8txc1+unjJsdmEgoqno3sCTFZt3Cw4eQ8Nwo=";
+  };
+
+  installPhase = ''
```
[the remark this answers is not in the document]
Comment (michaelCTS) [reply pkgs/tools/misc/ondir/default.nix:18]: What does this mean?

[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/tools/misc/ondir/default.nix:18]: Hi!

Please find some information here: https://discourse.nixos.org/t/what-does-runhook-do/13861

[the remark this answers is not in the document]
Comment (michaelCTS) [reply pkgs/tools/misc/ondir/default.nix:18]: Thanks. That should probably be an automatic check in the CI with an explanation. Pushed a fix

File: pkgs/tools/misc/ondir/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+}:
+
+stdenv.mkDerivation {
+  pname = "ondir";
+  version = "0.2.2";
```
Comment (drupol) [inline pkgs/tools/misc/ondir/default.nix:9]: Latest version, from 10 years ago is 0.2.3

**Review by drupol — CHANGES_REQUESTED** [review]
It's almost ok for me.

However, I'm curious as to your specific need for this derivation. Given that it hasn't been updated or maintained for a decade, I'm interested to understand your preference for it.

I'm not averse to merging this, but I am genuinely intrigued as to why you might require this, particularly when there are maintained alternatives like `direnv` that could potentially serve you better.

Could you shed some light on this for my own information?

**Comment by michaelCTS** [comment]
Hi @drupol

the reason I use this is that `direnv` doesn't support introducing functions in the running shell. 

For example, I have a node project that uses yarn. Were I to run the first-time setup commands (install my specific `yarn` version and the `yarn` dependencies) every time I entered the dev env with `nix-shell`, it would take a few seconds each time. For heavier projects like rust projects, I would only want to run the setup function once.

`direnv` evaluates `mkDerivation.shellHook` in a subshell however --> `_my_setup() {....}` is not available in the running shell.

I'm not entirely sure how `ondir` works, but it runs something like `eval $(ondir .....)` when you change directories and instead of doing some environmental extraction magic, I can just run `nix-shell` in the directory, have my setup function and whatever else I decide to put in the shell hook.

With `direnv` I'd have to work around this shortcoming

File: pkgs/tools/misc/ondir/default.nix (hunk 2 above)
Comment (michaelCTS) [reply pkgs/tools/misc/ondir/default.nix:9]: Thanks. Not sure how I missed that...

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Thanks for your first contribution ! Keep them coming :)

**Comment by michaelCTS** [comment]
Thank you for the review and for merging!

**Pull request merged**